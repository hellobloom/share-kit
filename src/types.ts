import Bowser from 'bowser'
import jwt from 'jsonwebtoken'
import {TAttestationTypeNames} from '@bloomprotocol/attestations-lib'

enum Action {
  attestation = 'request_attestation_data',
}

type BaseRequestData = {
  version: number
}

type RequestDataV1 = BaseRequestData & {
  version: 1
  action: Action
  token: string
  url: string
  payload_url: string
}

type RequestData = RequestDataV1

type BaseRequestDataPayload = {
  version: number
}

type DetailedAttestationTypeConfig = {
  name: string
  optional?: boolean
  completed_after?: string
  completed_before?: string
  provider_whitelist?: string[]
  provider_blacklist?: string[]
}

type RequestDataPayloadV1 = BaseRequestDataPayload & {
  version: 1
  org_logo_url: string
  org_name: string
  org_usage_policy_url: string
  org_privacy_policy_url: string
  types: (TAttestationTypeNames | DetailedAttestationTypeConfig)[]
}

type RequestDataPayload = RequestDataPayloadV1

type JsonWebTokenConfig = {
  token: string
  secretOrPublicKey: string | Buffer
  options?: jwt.VerifyOptions
}

enum ErrorCorrectionLevel {
  'L' = 1,
  'M' = 0,
  'Q' = 3,
  'H' = 2,
}

type QROptions = {
  ecLevel: keyof typeof ErrorCorrectionLevel
  size: number
  bgColor: string
  fgColor: string
  hideLogo: boolean
  padding: number
  logoImage?: string
  logoWidth?: number
  logoHeight?: number
  logoOpacity?: number
}

type SmallButtonType = 'square' | 'rounded-square' | 'circle' | 'squircle'

type MediumButtonType = 'log-in' | 'sign-up' | 'connect' | 'bloom' | 'verify'

type LargeButtonType = 'log-in' | 'sign-up' | 'connect' | 'bloom' | 'verify'

type ButtonSize = 'sm' | 'md' | 'lg'

type BaseButtonOptions = {
  callbackUrl: string
  size?: ButtonSize
}

type SmallButtonOptions = BaseButtonOptions & {
  size: 'sm'
  type: SmallButtonType
  invert?: boolean
}

type MediumButtonOptions = BaseButtonOptions & {
  size: 'md'
  type?: MediumButtonType
}

type LargeButtonOptions = BaseButtonOptions & {
  size?: 'lg'
  type?: LargeButtonType
}

type ButtonOptions = SmallButtonOptions | MediumButtonOptions | LargeButtonOptions

type ShouldRenderButton = (parsedResult: Bowser.Parser.ParsedResult) => boolean

type RequestElementResult = {
  update: (config: {jwtConfig: JsonWebTokenConfig; buttonOptions: ButtonOptions; qrOptions?: Partial<QROptions>}) => void
  remove: () => void
}

export {
  Action,
  RequestData,
  RequestDataPayload,
  JsonWebTokenConfig,
  ErrorCorrectionLevel,
  QROptions,
  SmallButtonType,
  MediumButtonType,
  LargeButtonType,
  ButtonSize,
  SmallButtonOptions,
  MediumButtonOptions,
  LargeButtonOptions,
  ButtonOptions,
  ShouldRenderButton,
  RequestElementResult,
}
