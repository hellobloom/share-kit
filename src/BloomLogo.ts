const logoWithColors = (config: {fgColor: string; bgColor: string}) =>
  `<svg width="25" height="25" viewBox="0 0 710 705" xmlns="http://www.w3.org/2000/svg">
    <rect x="0" y="0" width="710" height="705" fill="${config.bgColor}" />
    <g fill="${config.fgColor}">
      <g id="Leaf" transform="translate(260.000000, -27.000000)">
        <path d="M194.415739,215.192523 C194.258317,268.684765
          150.797287,312 97.2080838,312 C43.6188804,312
          0.157850908,268.684765 0.000429008653,215.192523
          C2.08721929e-14,215.192523 -3.50299401,197.114019
          15.1796407,156.874766 C47.118659,88.0834938
          95.9422811,58.317757 95.9422811,58.317757 C95.9422811,58.317757
          148.877246,91.5588785 180.404192,156.874766 C196.94399,191.141061
          194.416168,215.192523 194.416168,215.192523 Z" id="Combined-Shape" />
      </g>
      <g id="Leaf" transform="translate(357.500000, 576.000000) scale(1, -1)
        translate(-357.500000, -576.000000) translate(260.000000, 420.000000)">
        <path d="M194.415739,215.192523 C194.258317,268.684765
          150.797287,312 97.2080838,312 C43.6188804,312 0.157850908,268.684765
          0.000429008653,215.192523 C2.08721929e-14,215.192523 -3.50299401,197.114019
          15.1796407,156.874766 C47.118659,88.0834938 95.9422811,58.317757
          95.9422811,58.317757 C95.9422811,58.317757 148.877246,91.5588785
          180.404192,156.874766 C196.94399,191.141061 194.416168,215.192523
          194.416168,215.192523 Z" id="Combined-Shape" />
      </g>
      <g id="Leaf" transform="translate(563.000000, 352.500000) scale(1, -1)
        rotate(90.000000) translate(-563.000000, -352.500000)
        translate(465.500000, 196.500000)">
        <path d="M194.415739,215.192523 C194.258317,268.684765 150.797287,312
          97.2080838,312 C43.6188804,312 0.157850908,268.684765
          0.000429008653,215.192523 C2.08721929e-14,215.192523
          -3.50299401,197.114019 15.1796407,156.874766 C47.118659,88.0834938
          95.9422811,58.317757 95.9422811,58.317757 C95.9422811,58.317757
          148.877246,91.5588785 180.404192,156.874766 C196.94399,191.141061
          194.416168,215.192523 194.416168,215.192523 Z" id="Combined-Shape" />
      </g>
      <g id="Leaf" transform="translate(147.000000, 351.500000)
          scale(-1, -1) rotate(90.000000) translate(-147.000000, -351.500000)
          translate(49.500000, 195.500000)">
        <path d="M194.415739,215.192523 C194.258317,268.684765 150.797287,312
          97.2080838,312 C43.6188804,312 0.157850908,268.684765
          0.000429008653,215.192523 C2.08721929e-14,215.192523
          -3.50299401,197.114019 15.1796407,156.874766 C47.118659,88.0834938
          95.9422811,58.317757 95.9422811,58.317757 C95.9422811,58.317757
          148.877246,91.5588785 180.404192,156.874766 C196.94399,191.141061
          194.416168,215.192523 194.416168,215.192523 Z" id="Combined-Shape" />
      </g>
    </g>
  </svg>`

class BloomLogo {
  static getLogo = (config?: {fgColor?: string; bgColor?: string}) => {
    const defaultedConfig = Object.assign(
      {
        fgColor: '#6067f1',
        bgColor: '#fff',
      },
      config || {}
    )
    defaultedConfig.fgColor = encodeURIComponent(defaultedConfig.fgColor)
    defaultedConfig.bgColor = encodeURIComponent(defaultedConfig.bgColor)

    return `data:image/svg+xml;utf8,${logoWithColors(defaultedConfig)}`
  }
}

export {BloomLogo}
