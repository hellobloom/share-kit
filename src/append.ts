import extend from 'extend'

const url = require('url')

export const appendQuery = (uri: string, queryToAppend: {[key: string]: string | null}) => {
  const parts = url.parse(uri, true)
  const parsedQuery = extend(true, {}, parts.query, queryToAppend)

  const queryString = Object.keys(parsedQuery)
    .map(key => {
      const value = parsedQuery[key]
      return value === null ? encodeURIComponent(key) : encodeURIComponent(key) + '=' + encodeURIComponent(value)
    })
    .join('&')

  parts.query = null
  parts.search = queryString ? '?' + queryString : null

  return url.format(parts)
}
