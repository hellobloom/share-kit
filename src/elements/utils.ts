const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'

const generateId = () => {
  let rand = ''
  for (var i = 0; i < 4; i++) {
    rand += possible.charAt(Math.floor(Math.random() * possible.length))
  }

  return 'bloom-request-button-' + rand
}

export {generateId}
