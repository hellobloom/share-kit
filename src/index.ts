export * from './types'
export {renderRequestElement} from './renderRequestElement'
export {generateJWT} from './jwt'
