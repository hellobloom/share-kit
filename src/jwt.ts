import jwt from 'jsonwebtoken'

import {JsonWebTokenConfig, RequestData} from './types'

const verifyJWT = (config: JsonWebTokenConfig) => {
  jwt.verify(config.token, config.secretOrPublicKey, config.options)
}

const generateJWT = (data: RequestData, secretOrPrivateKey: jwt.Secret, options?: jwt.SignOptions) =>
  jwt.sign(data, secretOrPrivateKey, options)

export {verifyJWT, generateJWT}
