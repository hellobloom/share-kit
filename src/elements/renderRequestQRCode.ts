import {generateId} from './utils'
import {Options, RequestData, ErrorCorrectionLevel} from '../types'
import {BloomLogo} from '../BloomLogo'

const QRCodeImpl = require('qr.js/lib/QRCode')

const defaultOptions: Options = {
  hideLogo: false,
  ecLevel: 'L',
  size: 128,
  bgColor: '#fff',
  fgColor: '#6067f1',
  padding: 0,
}

type CellInfo = {
  top: number
  left: number
  size: number
  color: string
}

const makeDot = (ctx: CanvasRenderingContext2D, info: CellInfo) => {
  const centerX = info.left + info.size / 2
  const centerY = info.top + info.size / 2
  const radius = (info.size / 2) * 0.85

  ctx.save()
  ctx.beginPath()

  ctx.fillStyle = info.color

  ctx.arc(centerX, centerY, radius, 0, 2 * Math.PI, false)
  ctx.fill()

  ctx.closePath()
  ctx.restore()
}

const makeRect = (ctx: CanvasRenderingContext2D, color: string, rect: {x: number; y: number; w: number; h: number}) => {
  ctx.save()
  ctx.beginPath()

  ctx.fillStyle = color
  ctx.strokeStyle = color
  ctx.lineWidth = 1
  ctx.lineJoin = 'round'
  ctx.lineCap = 'round'

  ctx.rect(rect.x, rect.y, rect.w, rect.h)
  ctx.fill()
  ctx.stroke()

  ctx.closePath()
  ctx.restore()
}

enum CornerType {
  none,
  topLeft,
  topRight,
  bottomRight,
  bottomLeft,
}

const makeCorner = (ctx: CanvasRenderingContext2D, info: CellInfo, type: CornerType, part: EyePart) => {
  const outerRadius = part === EyePart.outer ? info.size : info.size / 4
  const innerRadius = outerRadius / 2

  const {top, left} = info
  const bottom = top + info.size
  const right = left + info.size
  const halfSize = info.size / 2

  ctx.save()
  ctx.beginPath()

  ctx.fillStyle = info.color
  ctx.strokeStyle = info.color
  ctx.lineWidth = 1
  ctx.lineJoin = 'round'
  ctx.lineCap = 'round'

  switch (type) {
    case CornerType.topLeft:
      ctx.moveTo(left, bottom)
      ctx.arcTo(left, top, right, top, outerRadius)
      ctx.lineTo(right, top)
      ctx.lineTo(right, bottom)
      ctx.lineTo(right + halfSize, bottom)
      ctx.arcTo(right, bottom, right, bottom + 5, innerRadius)
      ctx.lineTo(right, bottom + 5)
      ctx.lineTo(right, bottom)
      ctx.lineTo(left, bottom)
      break
    case CornerType.topRight:
      ctx.moveTo(left, bottom)
      ctx.lineTo(left, top)
      ctx.arcTo(right, top, right, bottom, outerRadius)
      ctx.lineTo(right, bottom)
      ctx.lineTo(left, bottom)
      ctx.lineTo(left, bottom + halfSize)
      ctx.arcTo(left, bottom, left - halfSize, bottom, innerRadius)
      ctx.lineTo(left - halfSize, bottom)
      ctx.lineTo(left, bottom)
      break
    case CornerType.bottomRight:
      ctx.moveTo(left, bottom)
      ctx.lineTo(left, top)
      ctx.lineTo(left - halfSize, top)
      ctx.arcTo(left, top, left, top - halfSize, innerRadius)
      ctx.lineTo(left, top - halfSize)
      ctx.lineTo(left, top)
      ctx.lineTo(right, top)
      ctx.arcTo(right, bottom, left, bottom, outerRadius)
      ctx.lineTo(left, bottom)
      break
    case CornerType.bottomLeft:
      ctx.moveTo(left, top)
      ctx.lineTo(right, top)
      ctx.lineTo(right, top - halfSize)
      ctx.arcTo(right, top, right + halfSize, top, innerRadius)
      ctx.lineTo(right + halfSize, top)
      ctx.lineTo(right, top)
      ctx.lineTo(right, bottom)
      ctx.arcTo(left, bottom, left, top, outerRadius)
      ctx.lineTo(left, top)
      break
    default:
      break
  }

  ctx.fill()
  ctx.stroke()

  ctx.closePath()
  ctx.restore()
}

class ConnectionType {
  static readonly none = 0
  static readonly left = 1 << 0
  static readonly right = 1 << 1
  static readonly top = 1 << 2
  static readonly bottom = 1 << 4
}

enum EyePart {
  inner,
  outer,
}

const makeEyeBit = (ctx: CanvasRenderingContext2D, info: CellInfo, connectionType: number, part: EyePart) => {
  const centerX = info.left + info.size / 2
  const centerY = info.top + info.size / 2
  const halfSize = info.size / 2

  const isLeft = (connectionType & ConnectionType.left) === ConnectionType.left
  const isRight = (connectionType & ConnectionType.right) === ConnectionType.right
  const isTop = (connectionType & ConnectionType.top) === ConnectionType.top
  const isBottom = (connectionType & ConnectionType.bottom) === ConnectionType.bottom

  if (isLeft !== isRight && isTop !== isBottom) {
    let cornerType = CornerType.none

    if (isBottom && isRight) {
      cornerType = CornerType.topRight
    } else if (isBottom && isLeft) {
      cornerType = CornerType.topLeft
    } else if (isTop && isRight) {
      cornerType = CornerType.bottomRight
    } else if (isTop && isLeft) {
      cornerType = CornerType.bottomLeft
    }

    makeCorner(ctx, info, cornerType, part)
  } else {
    if (isLeft) {
      makeRect(ctx, info.color, {x: centerX, y: info.top, w: halfSize, h: info.size})
    }

    if (isRight) {
      makeRect(ctx, info.color, {x: info.left, y: info.top, w: halfSize, h: info.size})
    }

    if (isTop) {
      makeRect(ctx, info.color, {x: info.left, y: info.top, w: info.size, h: halfSize})
    }

    if (isBottom) {
      makeRect(ctx, info.color, {x: info.left, y: centerY, w: info.size, h: halfSize})
    }
  }
}

const drawCanvas = (canvas: HTMLCanvasElement, data: RequestData, options: Partial<Options>) => {
  const defaultedOptions = {...defaultOptions, ...options}

  const {ecLevel, size, bgColor, fgColor, padding} = defaultedOptions

  const qr = new QRCodeImpl(-1, ErrorCorrectionLevel[ecLevel])
  qr.addData(JSON.stringify(data))
  qr.make()

  const ctx = canvas.getContext('2d')!
  ctx.clearRect(0, 0, canvas.width, canvas.height)

  const scale = window.devicePixelRatio || 1
  const cells: [boolean[]] = qr.modules
  const cellSize = size / cells.length
  // Add 4 to account for:
  // - The 1 extra on top and left of each cell
  // - The 1 extra from the border on top/left
  canvas.height = canvas.width = size * scale + 4
  canvas.style.height = canvas.style.width = `${size}px`
  canvas.style.padding = (100 * padding) / size + '%'
  ctx.scale(scale, scale)

  makeRect(ctx, bgColor, {x: 0, y: 0, w: size, h: size})

  cells.forEach((row: boolean[], rowIndex: number) => {
    row.forEach((cell: boolean, cellIndex: number) => {
      // This is a data bit
      if (cell) {
        const isTopLeftEye = cellIndex <= 7 && rowIndex <= 7
        const isTopRightEye = cellIndex >= cells.length - 7 && rowIndex <= 7
        const isBottomLeftEye = cellIndex <= 7 && rowIndex >= row.length - 7
        const isEye = isTopLeftEye || isTopRightEye || isBottomLeftEye

        // Add an extra 1 to the top/left so the border isn't cut off by the edge of the canvas
        const cellInfo = {
          color: fgColor,
          left: cellIndex * cellSize + 1,
          top: rowIndex * cellSize + 1,
          size: cellSize,
        }

        // Round the edges of eye bits
        if (isEye) {
          let connectionType = ConnectionType.none

          if (rowIndex > 0 && cells[rowIndex - 1][cellIndex]) {
            connectionType = connectionType | ConnectionType.top
          }

          if (rowIndex < cells.length - 1 && cells[rowIndex + 1][cellIndex]) {
            connectionType = connectionType | ConnectionType.bottom
          }

          if (cellIndex > 0 && cells[rowIndex][cellIndex - 1]) {
            connectionType = connectionType | ConnectionType.right
          }

          if (cellIndex < row.length - 1 && cells[rowIndex][cellIndex + 1]) {
            connectionType = connectionType | ConnectionType.left
          }

          const isInnerEyeX =
            (cellIndex >= 2 && cellIndex <= 4) || (cellIndex >= cells.length - 7 && cellIndex <= cells.length)
          const isInnerEyeY =
            (rowIndex >= 2 && rowIndex <= 4) || (rowIndex >= cells.length - 5 && rowIndex <= cells.length - 3)
          const isInnerEye = isInnerEyeX && isInnerEyeY

          const eyePart = isInnerEye ? EyePart.inner : EyePart.outer

          makeEyeBit(ctx, cellInfo, connectionType, eyePart)
        } else {
          makeDot(ctx, cellInfo)
        }
      }
    })
  })

  // If hideLogo is true then don't render any logo
  // If logoImage is not set default to BloomLogo with colors matching the rest of the QR code
  // Otherwise display the provided logo
  if (!options.hideLogo) {
    const logoImage =
      options.logoImage === undefined ? BloomLogo.getLogo({fgColor: fgColor, bgColor: bgColor}) : options.logoImage

    const image = new Image()
    image.onload = () => {
      const numberOfCellsToCover = Math.floor(cells.length * 0.2)
      const addExtra = numberOfCellsToCover % 2 === 0
      const defaultWidth = numberOfCellsToCover * cellSize + (addExtra ? cellSize : 0)
      const dwidth = options.logoWidth || defaultWidth
      const dheight = options.logoHeight || dwidth
      // Add 1 to accomodate for the 1 shift of the cells
      const dx = (size - dwidth) / 2 + 1
      const dy = (size - dheight) / 2 + 1
      image.width = dwidth
      image.height = dheight

      ctx.save()
      ctx.globalAlpha = options.logoOpacity || 1
      ctx.drawImage(image, dx, dy, dwidth, dheight)
      ctx.restore()
    }
    image.src = logoImage
  }
}

const renderRequestQRCode = (container: HTMLElement, data: RequestData, options: Partial<Options>) => {
  const id = generateId()

  const canvas = document.createElement('canvas')
  canvas.id = id

  drawCanvas(canvas, data, options)

  container.append(canvas)

  return {
    update: updateRequestQRCode(id),
    remove: removeRequestQRCode(id),
  }
}

const updateRequestQRCode = (id: string) => (data: RequestData, options: Partial<Options>) => {
  const canvas = document.querySelector<HTMLCanvasElement>(`#${id}`)

  if (!canvas) return

  drawCanvas(canvas, data, options)
}

const removeRequestQRCode = (id: string) => () => {
  const canvas = document.querySelector(`#${id}`)

  if (canvas) canvas.remove()
}

export {renderRequestQRCode}
