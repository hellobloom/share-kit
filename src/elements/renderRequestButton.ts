import {generateId} from './utils'
import {renderSmallRequestButton} from './buttons/renderSmallRequestButton'
import {renderMediumRequestButton} from './buttons/renderMediumRequestButton'
import {renderLargeRequestButton} from './buttons/renderLargeRequestButton'
import {verifyJWT} from '../jwt'
import {RequestElementResult, ButtonOptions, JsonWebTokenConfig} from '../types'

const getLink = (jwtConfig: JsonWebTokenConfig, callbackUrl: string) =>
  `https://bloom.co/download?request=${window.btoa(JSON.stringify(jwtConfig))}&callback-url=${encodeURIComponent(callbackUrl)}`

const render = (id: string, anchor: HTMLAnchorElement, config: {jwtConfig: JsonWebTokenConfig; buttonOptions: ButtonOptions}) => {
  verifyJWT(config.jwtConfig)

  anchor.href = getLink(config.jwtConfig, config.buttonOptions.callbackUrl)

  // Clear all children
  while (anchor.lastChild) {
    anchor.removeChild(anchor.lastChild)
  }

  const buttonOptions = config.buttonOptions

  if (buttonOptions.size === 'sm') {
    renderSmallRequestButton(id, anchor, buttonOptions.type, buttonOptions.invert)
  } else if (buttonOptions.size === 'md') {
    renderMediumRequestButton(id, anchor, buttonOptions.type || 'verify')
  } else {
    renderLargeRequestButton(id, anchor, buttonOptions.type || 'verify')
  }
}

const updateRequestButton = (id: string, container: HTMLElement) => (config: {
  jwtConfig: JsonWebTokenConfig
  buttonOptions: ButtonOptions
}) => {
  const anchor = container.querySelector<HTMLAnchorElement>(`#${id}`)

  if (!anchor) return

  render(id, anchor, config)
}

const removeRequestButton = (id: string, container: HTMLElement) => () => {
  const anchor = container.querySelector(`#${id}`)

  if (anchor) anchor.remove()
}

const renderRequestButton = (config: {
  container: HTMLElement
  jwtConfig: JsonWebTokenConfig
  buttonOptions: ButtonOptions
}): RequestElementResult => {
  const id = generateId()

  const anchor = document.createElement('a')
  anchor.id = id

  anchor.href = getLink(config.jwtConfig, config.buttonOptions.callbackUrl)
  anchor.target = '_blank'
  anchor.rel = 'norefferer noopener'

  render(id, anchor, config)

  config.container.appendChild(anchor)

  return {
    update: updateRequestButton(id, config.container),
    remove: removeRequestButton(id, config.container),
  }
}

export {renderRequestButton}
