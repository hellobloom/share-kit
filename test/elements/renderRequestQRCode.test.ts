import {renderRequestQRCode} from '../../src/elements/renderRequestQRCode'
import {QROptions, ButtonOptions, JsonWebTokenConfig, Action} from '../../src/types'
import {generateJWT} from '../../src/jwt'
import {publicKey, privateKey} from '../keys'

jest.mock('../../src/elements/utils', () => {
  return {
    ...jest.requireActual('../../src/elements/utils'),
    generateId: () => 'bloom-request-element-',
  }
})

describe('renderRequestQRCode', () => {
  let requestQRCode: {
    update: (config: {jwtConfig: JsonWebTokenConfig; buttonOptions: ButtonOptions; options?: Partial<QROptions>}) => void
    remove: () => void
  }
  let container: HTMLDivElement

  beforeAll(() => {
    container = document.createElement('div')
  })

  afterAll(() => {
    container.remove()
  })

  beforeEach(() => {
    requestQRCode = renderRequestQRCode({
      container,
      jwtConfig: {
        token: generateJWT(
          {
            version: 1,
            action: Action.attestation,
            token: 'token',
            url: 'https://receive-kit.bloom.co/api/receive',
            payload_url: 'https://receive-kit.bloom.co/api/get-payload',
          },
          privateKey,
          {
            algorithm: 'RS512',
          },
        ),
        secretOrPublicKey: publicKey,
        options: {
          algorithms: ['RS512'],
        },
      },
    })
  })

  afterEach(() => {
    requestQRCode.remove()
  })

  test('renders the qr code', () => {
    expect(container.innerHTML).toMatchSnapshot()
  })

  test('throws when the JWT is invalid', () => {
    expect(() => {
      requestQRCode.update({
        jwtConfig: {
          token: generateJWT(
            {
              version: 1,
              action: Action.attestation,
              token: 'token',
              url: 'https://receive-kit.bloom.co/api/receive',
              payload_url: 'https://receive-kit.bloom.co/api/get-payload',
            },
            privateKey,
            {
              algorithm: 'RS512',
            },
          ),
          secretOrPublicKey: 'invalid',
          options: {
            algorithms: ['RS512'],
          },
        },
        buttonOptions: {callbackUrl: 'https://bloom.co/callback-url'},
      })
    }).toThrowErrorMatchingSnapshot()
  })

  test('removes qr code', () => {
    requestQRCode.remove()

    expect(container.querySelector('canvas')).toBeNull()
  })
})
