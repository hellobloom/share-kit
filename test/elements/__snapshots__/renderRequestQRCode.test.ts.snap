// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderRequestQRCode renders the qr code 1`] = `"<canvas id=\\"bloom-request-element-\\" width=\\"132\\" height=\\"132\\" style=\\"width: 128px; height: 128px; padding: 0px; box-sizing: border-box; background-color: rgb(255, 255, 255);\\"></canvas>"`;

exports[`renderRequestQRCode throws when the JWT is invalid 1`] = `"error:0906D06C:PEM routines:PEM_read_bio:no start line"`;
